File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
import heapq
from collections import defaultdict

class Solution:
    def findCheapestPrice(self, n: int, flights: list[list[int]], src: int, dst: int, k: int) -> int:
        edges = defaultdict(list)
        visit = [0 for _ in range(n)]

        for s, e, w in flights:
            edges[s].append((w, e))

        h = [(0, src, k)]
        visit[src] += 1
        while h:
            weight, now, k = heapq.heappop(h)
            if now == dst:
                return weight
            if k >= 0:
                for w, e in edges[now]:
                    if visit[e] <= n:
                        heapq.heappush(h, (weight + w, e, k-1))
            visit[now] += 1

        return -1

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.rs
use std::{collections::{BinaryHeap}, cmp};

impl Solution {
    pub fn find_cheapest_price(n: i32, flights: Vec<Vec<i32>>, src: i32, dst: i32, k: i32) -> i32 {
        fn dij(graph: &Vec<Vec<(usize, isize)>>, start: usize, end: usize, k: usize) -> i32 {
            let mut h: BinaryHeap<(isize, usize, usize)> = BinaryHeap::new();
            let n = graph.len();
            
            h.push((0, start, 0));
            
            let mut dist: Vec<Vec<isize>> = vec![vec![isize::MAX; k + 2]; n];
            let mut min = isize::MAX;
    
            while let Some((cost, now, count)) = h.pop() {
                let cost = -cost;
                if now == end {
                    min = cmp::min(min, cost);
                    continue
                }
                
                if dist[now][count] <= cost {
                    continue
                }
                dist[now][count] = cost;
                for (next, c) in graph[now].iter() {
                    if count > k {
                        continue;
                    }
                    if cost + *c >= dist[*next][count + 1] {
                        continue;
                    }
                    h.push((-(cost + *c), *next, count + 1));
                }
            }
            if min == isize::MAX {
                return -1;
            }
            return min as i32;
        }
        
        let n = n as usize;
        let mut graph: Vec<Vec<(usize, isize)>> = vec![vec![]; n + 1];
        
        for flight in flights {
            let (start, end, cost) = (flight[0] as usize, flight[1] as usize, flight[2] as isize);
            graph[start].push((end, cost));
        }
        
        return dij(&graph, src as usize, dst as usize, k as usize);
    }
}